File: Math/p202_happy_number.py
class Solution:
    def isHappy(self, n: int) -> bool:
        
        def sumdigits(n):
            total=0
            while n>0:
                total+=(n%10)**2
                n//=10
            return total
        
        seen=set()

        while True:
            if n in seen:
                return False
            if n==1:
                return True
            seen.add(n)
            n= sumdigits(n)
        
#time complexity O(log n)
#space complexity O(log n)

File: Hashmap/p202_happy_number.cpp
class Solution {
public:
  bool isHappy(int n) {
    unordered_set<int> seen;
    while (true) {
      if (seen.count(n)) {
        return false;
      }
      if (n == 1) {
        return true;
      }
      seen.insert(n);
      n = getDigSum(n);
    }
  }

  int getDigSum(int n) {
    int total = 0;
    while (n > 0) {
      total += (n % 10) * (n % 10);
      n = n / 10;
    }
    return total;
  }
};
// time complexity O(log N)
// space complexity O(log N)